File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.java
class Solution {
    int n ;

    private int solve (String s, int l , int r, int[][] dp) {
        if(l > r) return 0;
        if(l == r) return 1;
        if(dp[l][r] != -1) return dp[l][r];

        if(s.charAt(l) == s.charAt(r)) {
            return dp[l][r] = 2 + solve(s,l+1,r-1,dp);
        }

        return dp[l][r] = Math.max(solve(s,l+1,r,dp),solve(s,l,r-1,dp));       
    }

    public int longestPalindromeSubseq(String s) {
        n = s.length();
        int[][] dp = new int[n][n];
        for(int i = 0; i<n; i++) {
            Arrays.fill(dp[i],-1);
        }
        return solve(s,0,n-1,dp);
    }
}

File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.cpp
class Solution {
public:
    int solve(string &s, int idx1, int idx2, vector<vector<int>> &dp) {
        if(idx1 > idx2) return 0;
        if(idx1 == idx2) return 1;
        if(dp[idx1][idx2] != -1) return dp[idx1][idx2];
        
        if(s[idx1] == s[idx2]) {
            return dp[idx1][idx2] = 2 + solve(s,idx1+1,idx2-1,dp);
        } else {
            return dp[idx1][idx2] = max(solve(s,idx1+1,idx2,dp),solve(s,idx1,idx2-1,dp));
        }
    }
    int longestPalindromeSubseq(string s) {
        int n = s.length();
        vector<vector<int>> dp (n, vector<int> (n,-1));
        return solve(s,0,n-1,dp);
    }
};